overlord/restart/restart.go

```diff
@@ -362,3 +362,134 @@ func ReplaceBootID(st *state.State, bootID string) {
 	rm := restartManager(st, "internal error: cannot mock a restart request before RestartManager initialization")
 	rm.bootID = bootID
 }
+
+// markTaskForRestart sets certain properties on a task to mark it for system restart.
+// The status argument is the status that the task will have after the system restart.
+func markTaskForRestart(t *state.Task, status state.Status) {
+	// Preserve previous restart behaviour for classic in the undo cases.
+	if release.OnClassic && (status == state.UndoStatus || status == state.UndoneStatus) {
+		t.SetStatus(status)
+		t.Logf("Skipped automatic system restart on classic system when undoing changes back to previous state")
+		return
+	}
+
+	rm := restartManager(t.State(), "internal error: cannot request a restart before RestartManager initialization")
+
+	// store current boot id to be able to check later if we have rebooted or not
+	t.Set("wait-for-system-restart-from-boot-id", rm.bootID)
+	t.SetToWait(status)
+	setWaitForSystemRestart(t.Change())
+
+	t.Logf("Task set to wait until a system restart allows to continue")
+}
+
+func changeRestartContext(chg *state.Change) (*RestartContext, error) {
```

Comment:
this merits a comment that if just created the returned struct will need initializing

Reply:
makes sense, added